File: tests/fixtures/c/interop_win_shm.c
/*
 * interop_win_shm.c - Server/client for cross-language Windows SHM interop.
 *
 * Usage:
 *   interop_win_shm server <run_dir> <service_name>
 *     Creates SHM region, receives 1 message, echoes it, exits.
 *
 *   interop_win_shm client <run_dir> <service_name>
 *     Attaches to SHM, sends 1 message, verifies echo, exits 0 on success.
 */

#ifdef _WIN32

#include "netipc/netipc_win_shm.h"
#include "netipc/netipc_named_pipe.h"
#include "netipc/netipc_protocol.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <windows.h>

#define AUTH_TOKEN 0xDEADBEEFCAFEBABEull
#define PROFILE   NIPC_WIN_SHM_PROFILE_HYBRID

static int run_server(const char *run_dir, const char *service)
{
    nipc_win_shm_ctx_t ctx;
    nipc_win_shm_error_t err = nipc_win_shm_server_create(
        run_dir, service, AUTH_TOKEN, PROFILE, 65536, 65536, &ctx);
    if (err != NIPC_WIN_SHM_OK) {
        fprintf(stderr, "server: shm create failed: %d\n", err);
        return 1;
    }

    /* Signal readiness */
    printf("READY\n");
    fflush(stdout);

    /* Receive one message */
    uint8_t msg[65536];
    size_t msg_len;
    err = nipc_win_shm_receive(&ctx, msg, sizeof(msg), &msg_len, 10000);
    if (err != NIPC_WIN_SHM_OK) {
        fprintf(stderr, "server: receive failed: %d\n", err);
        nipc_win_shm_destroy(&ctx);
        return 1;
    }

    /* Echo as response: decode header, flip kind, send back */
    if (msg_len >= NIPC_HEADER_LEN) {
        nipc_header_t hdr;
        nipc_header_decode(msg, msg_len, &hdr);
        hdr.kind = NIPC_KIND_RESPONSE;
        hdr.transport_status = NIPC_STATUS_OK;

        size_t payload_len = msg_len - NIPC_HEADER_LEN;
        size_t resp_len = NIPC_HEADER_LEN + payload_len;
        uint8_t *resp_buf = malloc(resp_len);
        if (resp_buf) {
            nipc_header_encode(&hdr, resp_buf, NIPC_HEADER_LEN);
            if (payload_len > 0)
                memcpy(resp_buf + NIPC_HEADER_LEN,
                       msg + NIPC_HEADER_LEN,
                       payload_len);

            err = nipc_win_shm_send(&ctx, resp_buf, resp_len);
            free(resp_buf);
            if (err != NIPC_WIN_SHM_OK) {
                fprintf(stderr, "server: send failed: %d\n", err);
                nipc_win_shm_destroy(&ctx);
                return 1;
            }
        }
    }

    nipc_win_shm_destroy(&ctx);
    return 0;
}

static int run_client(const char *run_dir, const char *service)
{
    nipc_win_shm_ctx_t ctx;

    /* Retry attach -- server may not be fully ready yet. */
    nipc_win_shm_error_t err = NIPC_WIN_SHM_ERR_OPEN_MAPPING;
    for (int i = 0; i < 500; i++) {
        err = nipc_win_shm_client_attach(run_dir, service, AUTH_TOKEN,
                                          PROFILE, &ctx);
        if (err == NIPC_WIN_SHM_OK)
            break;
        if (err == NIPC_WIN_SHM_ERR_OPEN_MAPPING ||
            err == NIPC_WIN_SHM_ERR_BAD_MAGIC)
            Sleep(10);
        else
            break;
    }
    if (err != NIPC_WIN_SHM_OK) {
        fprintf(stderr, "client: attach failed: %d\n", err);
        return 1;
    }

    /* Build a payload with known pattern */
    uint8_t payload[256];
    for (int i = 0; i < (int)sizeof(payload); i++)
        payload[i] = (uint8_t)(i & 0xFF);

    nipc_header_t hdr = {
        .magic       = NIPC_MAGIC_MSG,
        .version     = NIPC_VERSION,
        .header_len  = NIPC_HEADER_LEN,
        .kind        = NIPC_KIND_REQUEST,
        .code        = NIPC_METHOD_INCREMENT,
        .flags       = 0,
        .item_count  = 1,
        .message_id  = 12345,
        .payload_len = sizeof(payload),
        .transport_status = NIPC_STATUS_OK,
    };

    /* Build complete message */
    uint8_t msg_buf[NIPC_HEADER_LEN + 256];
    nipc_header_encode(&hdr, msg_buf, NIPC_HEADER_LEN);
    memcpy(msg_buf + NIPC_HEADER_LEN, payload, sizeof(payload));

    err = nipc_win_shm_send(&ctx, msg_buf, sizeof(msg_buf));
    if (err != NIPC_WIN_SHM_OK) {
        fprintf(stderr, "client: send failed: %d\n", err);
        nipc_win_shm_close(&ctx);
        return 1;
    }

    /* Receive response */
    uint8_t resp[65536];
    size_t resp_len;
    err = nipc_win_shm_receive(&ctx, resp, sizeof(resp), &resp_len, 10000);
    if (err != NIPC_WIN_SHM_OK) {
        fprintf(stderr, "client: receive failed: %d\n", err);
        nipc_win_shm_close(&ctx);
        return 1;
    }

    /* Verify */
    int ok = 1;
    if (resp_len < NIPC_HEADER_LEN) {
        fprintf(stderr, "client: response too short\n");
        ok = 0;
    } else {
        nipc_header_t rhdr;
        nipc_header_decode(resp, resp_len, &rhdr);

        if (rhdr.kind != NIPC_KIND_RESPONSE) {
            fprintf(stderr, "client: expected RESPONSE, got %u\n", rhdr.kind);
            ok = 0;
        }
        if (rhdr.message_id != 12345) {
            fprintf(stderr, "client: expected message_id 12345, got %llu\n",
                    (unsigned long long)rhdr.message_id);
            ok = 0;
        }
        size_t rpayload_len = resp_len - NIPC_HEADER_LEN;
        if (rpayload_len != sizeof(payload)) {
            fprintf(stderr, "client: payload length mismatch: %zu vs %zu\n",
                    rpayload_len, sizeof(payload));
            ok = 0;
        }
        if (rpayload_len == sizeof(payload) &&
            memcmp(resp + NIPC_HEADER_LEN, payload, sizeof(payload)) != 0) {
            fprintf(stderr, "client: payload data mismatch\n");
            ok = 0;
        }
    }

    nipc_win_shm_close(&ctx);

    if (ok)
        printf("PASS\n");
    else
        printf("FAIL\n");

    return ok ? 0 : 1;
}

int main(int argc, char **argv)
{
    if (argc != 4) {
        fprintf(stderr, "Usage: %s <server|client> <run_dir> <service_name>\n", argv[0]);
        return 1;
    }

    const char *mode = argv[1];
    const char *run_dir = argv[2];
    const char *service = argv[3];

    if (strcmp(mode, "server") == 0)
        return run_server(run_dir, service);
    else if (strcmp(mode, "client") == 0)
        return run_client(run_dir, service);
    else {
        fprintf(stderr, "Unknown mode: %s\n", mode);
        return 1;
    }
}

#else /* !_WIN32 */

#include <stdio.h>
int main(int argc, char **argv)
{
    (void)argc; (void)argv;
    printf("Windows SHM interop not supported on this platform\n");
    return 1;
}

#endif /* _WIN32 */

File: tests/fixtures/rust/src/bin/interop_win_shm.rs
//! Simple server/client for cross-language Windows SHM interop tests.
//!
//! Usage:
//!   interop_win_shm server <run_dir> <service_name>
//!     Creates SHM region, receives 1 message, echoes it, exits.
//!
//!   interop_win_shm client <run_dir> <service_name>
//!     Attaches to SHM, sends 1 message, verifies echo, exits 0 on success.

#[cfg(windows)]
use netipc::protocol::{self, Header};
#[cfg(windows)]
use netipc::transport::win_shm::{WinShmContext, PROFILE_HYBRID};

#[cfg(windows)]
const AUTH_TOKEN: u64 = 0xDEADBEEFCAFEBABE;

#[cfg(windows)]
fn build_message(kind: u16, code: u16, message_id: u64, payload: &[u8]) -> Vec<u8> {
    let hdr = Header {
        magic: protocol::MAGIC_MSG,
        version: protocol::VERSION,
        header_len: protocol::HEADER_LEN,
        kind,
        code,
        flags: 0,
        transport_status: protocol::STATUS_OK,
        payload_len: payload.len() as u32,
        item_count: 1,
        message_id,
    };
    let mut buf = vec![0u8; protocol::HEADER_SIZE + payload.len()];
    hdr.encode(&mut buf[..protocol::HEADER_SIZE]);
    buf[protocol::HEADER_SIZE..].copy_from_slice(payload);
    buf
}

#[cfg(windows)]
fn run_server(run_dir: &str, service: &str) -> Result<(), Box<dyn std::error::Error>> {
    let mut ctx = WinShmContext::server_create(
        run_dir, service, AUTH_TOKEN, PROFILE_HYBRID, 65536, 65536,
    )?;

    // Signal readiness
    println!("READY");

    // Receive one message
    let mut buf = vec![0u8; 65536];
    let mlen = ctx.receive(&mut buf, 10000)?;

    if mlen < protocol::HEADER_SIZE {
        return Err("message too short".into());
    }

    // Parse and echo as response
    let hdr = Header::decode(&buf[..mlen])?;
    let payload_copy: Vec<u8> = buf[protocol::HEADER_SIZE..mlen].to_vec();
    let resp = build_message(
        protocol::KIND_RESPONSE,
        hdr.code,
        hdr.message_id,
        &payload_copy,
    );
    ctx.send(&resp)?;

    ctx.destroy();
    Ok(())
}

#[cfg(windows)]
fn run_client(run_dir: &str, service: &str) -> Result<(), Box<dyn std::error::Error>> {
    // Retry attach -- server may not be fully ready yet
    let mut ctx = None;
    for _ in 0..500 {
        match WinShmContext::client_attach(run_dir, service, AUTH_TOKEN, PROFILE_HYBRID) {
            Ok(c) => {
                ctx = Some(c);
                break;
            }
            Err(netipc::transport::win_shm::WinShmError::OpenMapping(_))
            | Err(netipc::transport::win_shm::WinShmError::BadMagic) => {
                std::thread::sleep(std::time::Duration::from_millis(10));
            }
            Err(e) => return Err(e.into()),
        }
    }
    let mut ctx = ctx.ok_or("client: attach failed after retries")?;

    // Build payload with known pattern
    let payload: Vec<u8> = (0..256).map(|i| (i & 0xFF) as u8).collect();
    let msg = build_message(
        protocol::KIND_REQUEST,
        protocol::METHOD_INCREMENT,
        12345,
        &payload,
    );
    ctx.send(&msg)?;

    // Receive response
    let mut resp_buf = vec![0u8; 65536];
    let rlen = ctx.receive(&mut resp_buf, 10000)?;
    let resp_copy: Vec<u8> = resp_buf[..rlen].to_vec();

    let mut ok = true;
    if resp_copy.len() < protocol::HEADER_SIZE {
        eprintln!("client: response too short");
        ok = false;
    } else {
        let rhdr = Header::decode(&resp_copy)?;
        if rhdr.kind != protocol::KIND_RESPONSE {
            eprintln!("client: expected RESPONSE, got {}", rhdr.kind);
            ok = false;
        }
        if rhdr.message_id != 12345 {
            eprintln!("client: expected message_id 12345, got {}", rhdr.message_id);
            ok = false;
        }
        let resp_payload = &resp_copy[protocol::HEADER_SIZE..];
        if resp_payload.len() != 256 {
            eprintln!(
                "client: payload length mismatch: {} vs 256",
                resp_payload.len()
            );
            ok = false;
        }
        if resp_payload != &payload[..] {
            eprintln!("client: payload data mismatch");
            ok = false;
        }
    }

    ctx.close();

    if ok {
        println!("PASS");
        Ok(())
    } else {
        println!("FAIL");
        std::process::exit(1);
    }
}

fn main() {
    #[cfg(not(windows))]
    {
        eprintln!("Windows SHM interop not supported on this platform");
        std::process::exit(1);
    }

    #[cfg(windows)]
    {
        let args: Vec<String> = std::env::args().collect();
        if args.len() != 4 {
            eprintln!(
                "Usage: {} <server|client> <run_dir> <service_name>",
                args[0]
            );
            std::process::exit(1);
        }

        let mode = &args[1];
        let run_dir = &args[2];
        let service = &args[3];

        let result = match mode.as_str() {
            "server" => run_server(run_dir, service),
            "client" => run_client(run_dir, service),
            _ => {
                eprintln!("Unknown mode: {mode}");
                std::process::exit(1);
            }
        };

        if let Err(e) = result {
            eprintln!("{mode}: {e}");
            std::process::exit(1);
        }
    }
}
